### DISCUSSION TITLE
Allow DAGs to be scheduled when any related dataset has been updated

### DISCUSSION THREAD

**Discussion by hugochinchilla** [apache/airflow#28253] [Ideas]
Currently the behavior for a DAG which depends on multiple DataSets is to be executed only when all datasets have been updated since the last DAG execution. This feature would enable a DAG to run every time any of the multiple DataSets change.

The motivation is to allow better traceability of changes in a derived dataset, being able to track down changes along all the DAGs chain for every change in every dataset.

I'm able to provide a PR for this feature if the feature is desired. If there are current plans for this I would like to hear from them as I want to align the proposal with the current plan, I haven't found anything in the issues or the mailing list, but the docs already make it implicit that this feature is planed somehow.

**Comment by potiuk** [comment]
cc: @Taragolis @ashb - any plans in the pipeline for that one?

**Reply by Taragolis** [reply]
@potiuk I thought you mention me by mistake 😃

**Reply by ashb** [reply]
No active plans but it's definitely on the feature list we want/had in mind when building datasets 

A PR would be great @hugochinchilla - but before we do we should have a discussion about what the syntax is for this going to be

**Reply by potiuk** [reply]
Oh yeah @uranusjr was suppoed to be :)

**Comment by hugochinchilla** [comment]
@ashb for the syntax I have some ideas:

Adding a mode argument which can be AND or OR, either using constants or an Enum class, the DAG argument could be named
`dataset_schedule_mode`

It could also be a boolean flag named `run_on_any_dataset_changed`  or something like that.

Finally we could allow to pass a scheduling object as schedule param where one can freely combine AND and OR relationships between datasets. It would look something like this `DAG(schedule=or(Dataset('a'), and(Dataset('b'), Dataset('c')), ...)`

I think the last idea is the most powerful one, but personally I would rule it out for a first iteration, as its not possible to do the resolution via plain SQL as it's done now and I would prefer not to have to change too much code of how this is currently implemented.

**Reply by potiuk** [reply]
Powerful features are also often confusing. The Power of Airflow also comes from the way we can visualise dependencies and I think the last idea is also most confusing one in that respect. 

One of the important features of datasets is that you can easily reason about them when you see them in a graph, and I can't really imagine an easy way to explain it whit a visualisation in the way that will be easy to understand if you start mixing logical conditions.

I believe it should be simple - either "all" or "any" - so basically `run_on_any_dataset_changed` (maybe differently named - naming is difficutl) -  I think that proposal should also include a proposal on how to (without actual graphcs but with general approach).

Also I think we do not really neeed the "powerful" one. Once you have "all" and "any", you can start combining datasets and dependnencies in the cascading way using EmptyOperator and existiing dependencies to achieve any logic you want:

```
DS1 - (ANY) -> EmptyOperator1
DS2 |                          \
DS3 /                            -   (DS1 OR  DS2 OR DS3) AND (DS4 OR DS5) Final Operator       
                               /
DS4 - (ANY) -> EmptyOperator2  
DS5 /
```

And if the "ANY" visualisation will be easily distinguishable from (ALL), this should harnes the power of Airflow's dependency visualisations to visualise the dependencies nicely.

**Reply by hugochinchilla** [reply]
Then I will start writing a PR using a toggle for each DAG, we can change the name of the argument when we come up with a better one.

**Comment by hugochinchilla** [comment]
PR already submitted https://github.com/apache/airflow/pull/28333
